config best practices doc modifications

amygdala opened the pull request:
This is an attempt to flesh out the 'config best practices' page some more.
The eventual goal is to cover a lot of this stuff in enough detail that it can almost be used as a checklist that we can point people to when they submit new examples.

amygdala committed:
config best practices doc edits

bgrant0607 commented:
Thanks. Did anything change other than formatting and a few references to examples?

bgrant0607 commented:
Note to @janetkuo: Everywhere we encourage users to use ReplicationController should change to advocate Deployment once it's beta.

amygdala commented:
@bgrant0607 In addition to reorg and formatting, I added a new little section on container image tagging (I've seen this cause issues for people), and included references to the Job and Deployment objects in the discussions (descriptions basically copied from elsewhere in the docs). 
I also added the last bullet under "using labels", on debugging.

janetkuo commented:
@amygdala Thanks for the fix. Since you changed the title of this document, can you update the other documents that link to it?
- docs/devel/coding-conventions.md
- docs/user-guide/managing-deployments.md
- docs/user-guide/README.md
- docs/user-guide/kubectl-overview.md

janetkuo commented on docs/user-guide/config-best-practices.md:
```diff
@@ -31,27 +31,131 @@ Documentation for other releases can be found at
+- Don't specify default values unnecessarily, in order to simplify and minimize configs, and to
+  reduce error. For example, omit the selector and labels in a `ReplicationController` if you want
+  them to be the same as the labels in its `podTemplate`, since those fields are populated from the
+  `podTemplate` labels by default. See the [guestbook app's](../../examples/guestbook/) .yaml files for some [examples](../../examples/guestbook/frontend-controller.yaml) of this.
```
../../examples/guestbook/frontend-controller.yaml should comment out the unnecessary default metadata.labels. Can you update it too?

amygdala replied:
@janetkuo Janet, on this one -- linzichang's recent update essentially did that (in head, might not be in a tagged release yet).  Were you thinking something different than he's got there?

janetkuo replied:
I meant https://github.com/kubernetes/kubernetes/blob/master/examples/guestbook/frontend-controller.yaml#L7-L9

amygdala replied:
sorry, missed that.  done.

janetkuo commented on docs/user-guide/config-best-practices.md:
```diff
@@ -31,27 +31,131 @@ Documentation for other releases can be found at
+
+  Controllers are almost always preferable to creating pods, except for some explicit
+  [`restartPolicy: Never`](pod-states.md#restartpolicy) scenarios.  A
+  [Job](jobs.md) object (currently in Beta), may also be appropriate. A _job_ creates one or more
```
This part contains too much jobs and too few replication controllers information. Just present enough information for why we should avoid naked pods and link to both documents for more details.

amygdala replied:
done.

janetkuo commented on docs/user-guide/config-best-practices.md:
```diff
@@ -31,27 +31,131 @@ Documentation for other releases can be found at
+
+- It's typically best to create a [service](services.md) before corresponding [replication
+  controllers](replication-controller.md), so that the scheduler can spread the pods comprising the
+  service. You can also create a replication controller without specifying replicas, create a
```
Perhaps we can mention that creating a rc without specifying replicas means its replicas=1.

amygdala replied:
done.

janetkuo commented on docs/user-guide/config-best-practices.md:
```diff
@@ -31,27 +31,131 @@ Documentation for other releases can be found at
+  service, then scale up the replication controller. This can be useful in ensuring that one replica works before creating lots of them.
+
+- Don't use `hostPort` (which specifies the port number to expose on the host) unless absolutely
+  necessary. When you bind a Pod to a `hostPort`, there are a limited number of places that pod can
```
Add `(e.g., for a node daemon)` after `unless absolutely necessary`.

amygdala replied:
done.

janetkuo commented on docs/user-guide/config-best-practices.md:
```diff
@@ -31,27 +31,131 @@ Documentation for other releases can be found at
+- Write your configuration files using YAML rather than JSON. They can be used interchangeably in almost all scenarios, but YAML tends to be more user-friendly for config.
+
+- Group related objects together in a single file where this makes sense. This format is often easier to manage than separate files. See the [guestbook-all-in-one.yaml](../../examples/guestbook/all-in-one/guestbook-all-in-one.yaml) file as an example of this syntax.
+(Note that you can also call `kubectl create` on a directory of config files— see below for more detail).
```
`kubectl create -f` isn't the only command that takes a directory of config files. Most of the kubectl commands do. Pointing out `kubectl create` may be misleading.

janetkuo commented:
Added a few comments.

amygdala commented:
I'll include updates to the docs that link to this shortly, thanks for catching that.

amygdala commented:
@janetkuo okay, PTAL. I did not need to change one of the files that links to this one.

amygdala commented:
@janetkuo Janet, just a ping on this if you are around this week.  It should be ready to go.  I assume these changes, since they're all docs changes, are not implicated in the Jenkins issue...

janetkuo commented:
Thanks! Mostly LG except for these two nits:
- Comment out the unnecessary default metadata.labels in https://github.com/kubernetes/kubernetes/blob/master/examples/guestbook/frontend-controller.yaml#L7-L9
- docs/user-guide/kubectl-overview.md: update the link of "Use YAML rather than JSON" to docs/user-guide/config-best-practices.md#general-config-tips

amygdala commented:
@janetkuo Thanks. The above changes made; PTAL again when you get a chance.

janetkuo commented:
@k8s-bot unit test this
